### PR TITLE
feature(rome_js_parser): coverage test suite for ts symbols file

### PR DESCRIPTION
**Pull request by xunilrj**
see https://github.com/rome/tools/issues/2488

This PR parses Microsoft/ts .symbols file;

#### Microsoft ts .symbols file

We are running only files without any errors.

```
symbols/msts: Ran 5946 tests in 19.05s
┌───────────┬────────┬────────┬────────┬──────────┐
│ Tests ran │ Passed │ Failed │ Panics │ Coverage │
├───────────┼────────┼────────┼────────┼──────────┤
│      5946 │   1613 │   4328 │      5 │    27.13 │
└───────────┴────────┴────────┴────────┴──────────┘
```

All ```parse_*``` functions are simplified versions of ```nom```, just to avoid including another crate.

#### Fake Implementation

```Symbol``` and ```symbols``` are fake and only exist to make the coverage test to compile and run.

### CODE REVIEW

**Comment by MichaReiser** [comment]
I would find it useful if you could write up some overall architecture where the work on symbols and scopes is going. It's otherwise difficult to review the PR from an architectural point and only allows me to point out coding issues. 

My understanding of the problem is and things that are important to know are:

## Symbols

* It should be possible to resolve the symbol for a node, e.g. by querying a semantic model
* A symbol allows to 
  * query all its declarations and accessing the nodes (e.g. in case of overloading)
  * query all its references/uses and accessing the nodes
  * querying its child symbols. For example, to get all members of a class. This raises the point that languages have different symbols. What representation do we use? Somethin similar to `SyntaxNode<L>` and `AstNode` or is `JsSymbol` an enum
  * querying the parent symbol if any

## Scope

I'm deliberately listing scopes independently from `Symbols` because there are different requirements to it and symbols also exist for things that aren't part of a scope. 

* It should be possible to get a node's scope. For example, to suggest another variable name in case of a re-declaration (or to fix a reference if the name was misspelt). 
* It should be possible to get all symbols from a scope. Useful when introducing a new variable to make sure there are no name clashes. 
* It should be possible to get the parent scope
* Maybe: It should be possible to get all child scopes

## Semantic Model

I see it as likely that we need some form of a semantic model that is the entry point for anyone querying such information. The main question here is how does the API look like and how do we intend to build it up (individual method calls or a single method that builds the whole semantic model). 

Having this outline will allow you to describe in the PR summary which problem you are solving and how it helps build the final solution. 

I think it would further help to draft an issue that splits this problem into more tangible sub-problems and outline the work involved for each, so that someone else could help with the work or step in in case you are absent.

**Comment by xunilrj** [comment]
> I would find it useful if you could write up some overall architecture where the work on symbols and scopes is going.

I created this discussion to address the high-level architecture (https://github.com/rome/tools/discussions/2603) and this issue (https://github.com/rome/tools/issues/2488) to track the implementation details on a lower level.

**Comment by xunilrj** [comment]
> What's the difference between `xtask/coverage/src/{mod.rs,msts.rs,utils.rs}` and `xtask/coverage/src/symbols/{mod.rs,msts.rs,utils.rs}` ? It looks like those files are duplicated

You are right. I probably copied and pasted it without realizing it. Fixed.

File: xtask/coverage/src/symbols/msts.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        let t = TestCaseFiles::single(self.name.clone(), code.clone(), SourceType::tsx());
+
+        let r = rome_js_parser::parse(&code, 0, SourceType::tsx());
+        let mut actual: Vec<_> = rome_js_parser::symbols::symbols(r.syntax())
+            .into_iter()
+            .filter(|x| !x.name().contains('\"') && !x.name().contains('\''))
+            .collect();
+        actual.sort_by_key(|x| x.range().start());
+
+        if std::env::var("PRINT_CMP").is_ok() {
```
Comment (ematipico) [inline xtask/coverage/src/symbols/msts.rs:58]: Where and when do we use this environment variable? Maybe we should update the `CONTRIBUTING.md` file?

File: xtask/coverage/src/symbols/msts.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                files: t,
+                errors: vec![],
+            }
+        } else {
+            for (expected, actual) in expected.symbols.iter().zip(actual) {
+                let are_names_eq = expected.name == actual.name();
+                // let are_paths_eq = expected.name == actual.name;
+                //TODO check decls
+                if !are_names_eq
+                /*|| !are_paths_eq*/
```
Comment (ematipico) [inline xtask/coverage/src/symbols/msts.rs:101]: To remove?

File: xtask/coverage/src/symbols/msts.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+        if expected.symbols.len() != actual.len() {
+            TestRunOutcome::IncorrectlyErrored {
+                files: t,
+                errors: vec![],
+            }
+        } else {
+            for (expected, actual) in expected.symbols.iter().zip(actual) {
+                let are_names_eq = expected.name == actual.name();
+                // let are_paths_eq = expected.name == actual.name;
```
Comment (ematipico) [inline xtask/coverage/src/symbols/msts.rs:98]: To remove?

File: xtask/coverage/src/symbols/msts.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                //TODO check decls
```
Comment (ematipico) [inline xtask/coverage/src/symbols/msts.rs:99]: Will this be done in another PR?

File: xtask/coverage/src/symbols/msts.rs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        input,
+        Decl {
+            file: file.to_string(),
+            row_start: row_start.parse().ok(),
+            col_start: col_start.parse().ok(),
+        },
+    ))
+}
+
+/// see xtask\coverage\Typescript\src\harness\typeWriter.ts
```
Comment (ematipico) [inline xtask/coverage/src/symbols/msts.rs:190]: Where's this file? Maybe it's best to write a better documentation? At least, for example what this function is doing on a high level?

File: xtask/coverage/src/symbols/utils.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    if qty > 0 {
+        Some((&input[qty..], &input[0..qty]))
+    } else {
+        None
+    }
+}
+
+pub fn parse_whitespace0(input: &'_ str) -> (&'_ str, &'_ str) {
```
Comment (ematipico) [inline xtask/coverage/src/symbols/utils.rs:25]: Why the `0` in the name?

**Review by ematipico — APPROVED** [review]

File: xtask/coverage/src/symbols/msts.rs (hunk 1 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs:58]: Only here.
I can rename it and document it.

File: xtask/coverage/src/symbols/msts.rs (hunk 2 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs:101]: These are more todos from future PRs. But I can remove them for now.

File: xtask/coverage/src/symbols/msts.rs (hunk 4 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs:99]: Yes.

File: xtask/coverage/src/symbols/utils.rs (hunk 6 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/utils.rs:25]: Because it accepts no (zero) whitespaces. It is following nom's convention: https://docs.rs/nom/latest/nom/character/complete/fn.space0.html

File: xtask/coverage/src/symbols/msts.rs (hunk 1 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs]: Instead of having an obscure environment variable that needs documentation, I changed the code to print to ```debug!```. And, of course, it can be configured normally with ```RUST_LOG```, if needed.

I think this makes more sense.

File: xtask/coverage/src/symbols/msts.rs (hunk 3 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs]: Done.

File: xtask/coverage/src/symbols/msts.rs (hunk 4 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs]: Done.

File: xtask/coverage/src/symbols/msts.rs (hunk 5 above)
Comment (xunilrj) [reply xtask/coverage/src/symbols/msts.rs]: Done.
